File: plugins/superheroes/lib/phase_step.py
# plugins/superheroes/lib/phase_step.py
"""Per-phase decision: (phase_result, gate) -> the single terminal the showrunner.js shell
forwards. Pure + fail-closed (the band's recover/ci_loop/loop_state pattern). The JS computes
nothing; this owns the per-phase judgement (FR-5/FR-6/FR-7/FR-8).

Ordering is the safety contract: the assumption / low-confidence parks are evaluated BEFORE
the gate, so a phase that records an assumption or low confidence parks even when its gate is
'passed'. Then the gate maps over the spec's recognized set {pending, passed, changes-requested};
any other or unreadable value fails closed to park_unexpected_gate.
"""


def decide(phase_result, gate):
    pr = phase_result or {}
    # 1. self-assessed park signals first (before the gate) — the safety ordering.
    if pr.get("assumptions"):
        # #212: name WHICH assumption(s) — the payload carries the list; a bare "recorded a material
        # assumption" hid the real cause. The infra parkReason override still wins at the consumer, so
        # this richer reason surfaces where no override was set (e.g. review-code sub-parks).
        detail = "; ".join(str(a) for a in pr.get("assumptions") or [])
        reason = "phase recorded a material assumption"
        if detail:
            reason += ": " + detail
        return {"action": "park_assumption", "reason": reason}
    if pr.get("confidence") == "low":
        return {"action": "park_low_confidence",
                "reason": "phase recorded confidence below the parking threshold"}
    # 2. gate dimension. None = an authoring phase with no review gate.
    if gate is None or gate == "passed":
        return {"action": "proceed",
                "reason": "no review gate" if gate is None else "gate passed"}
    if gate == "changes-requested":
        # #212: a review park must survive the phase layer — thread the named terminal reason the
        # review phase put on parkDetail (e.g. "cannot-certify: <seat> returned no verification
        # receipt after retry") so the workflow park reads it instead of the bare flatten.
        detail = pr.get("parkDetail")
        reason = "review requested changes"
        if detail:
            reason += " — " + str(detail)
        return {"action": "park_changes_requested", "reason": reason}
    if gate == "pending":
        return {"action": "park_pending",
                "reason": "gate not passed (pending / not yet approved)"}
    return {"action": "park_unexpected_gate",
            "reason": "unexpected or unreadable gate value: %r" % (gate,)}

File: plugins/superheroes/lib/phase_step.js
// plugins/superheroes/lib/phase_step.js
// Faithful JS twin of phase_step.py:decide — parity-locked. Safety ordering: assumption /
// low-confidence parks are evaluated BEFORE the gate (a recorded assumption parks even on a
// passed gate). Pure + fail-closed.
function pyReprStr(v) {
  // Python %r for a simple str: single-quoted, backslash- and quote-escaped.
  if (typeof v === 'string') return "'" + v.replace(/\\/g, '\\\\').replace(/'/g, "\\'") + "'"
  if (v === null || v === undefined) return 'None'
  return String(v)
}
function decide(phaseResult, gate) {
  const pr = phaseResult || {}
  if (pr.assumptions && pr.assumptions.length) {
    // #212: name WHICH assumption(s) — the payload carries the list. The infra parkReason override
    // still wins at the consumer; this richer reason surfaces where no override was set.
    const detail = pr.assumptions.map((a) => String(a)).join('; ')
    let reason = 'phase recorded a material assumption'
    if (detail) reason += ': ' + detail
    return { action: 'park_assumption', reason }
  }
  if (pr.confidence === 'low') {
    return { action: 'park_low_confidence', reason: 'phase recorded confidence below the parking threshold' }
  }
  if (gate === null || gate === undefined || gate === 'passed') {
    return { action: 'proceed', reason: (gate === null || gate === undefined) ? 'no review gate' : 'gate passed' }
  }
  if (gate === 'changes-requested') {
    // #212: thread the named terminal reason (parkDetail) so the workflow park survives the flatten.
    let reason = 'review requested changes'
    if (pr.parkDetail) reason += ' — ' + String(pr.parkDetail)
    return { action: 'park_changes_requested', reason }
  }
  if (gate === 'pending') return { action: 'park_pending', reason: 'gate not passed (pending / not yet approved)' }
  return { action: 'park_unexpected_gate', reason: 'unexpected or unreadable gate value: ' + pyReprStr(gate) }
}
module.exports = { decide }
